### PR TITLE
147 make the synthesis process reproducible

### PR DESCRIPTION
**Pull request by KoomenErina**
This pull requests makes the entire proces of making synthetic data reproducible.

### CODE REVIEW

File: src/synthpop/methods/tree_utils.py [hunk 1]
```diff
@@ -122,13 +124,16 @@ def fit_sampler(self, leaf_ids: npt.ArrayLike, y: npt.ArrayLike) -> Self:
                     "distributions and may be sampled. Review your input data "
                     "if this is unintended.")
 
-        if isinstance(self.random_state, np.random.Generator):
-            # Extract seed is not possible → treat as non-resettable (repeated calls do not give the same output)
-            self._seed = None
-            self.random_state_ = self.random_state
+        # if isinstance(self.random_state, np.random.Generator):
```
Comment (anneliesdej) [inline src/synthpop/methods/tree_utils.py]: Can these comments be removed or do you think they still add necessary information?

File: src/synthpop/methods/tree_utils.py [hunk 2]
```diff
@@ -159,12 +164,14 @@ def sample_from_leaves(self, leaf_ids: npt.ArrayLike) -> npt.NDArray:
...
-        if seed is not None:
-            rng = np.random.default_rng(seed)
-        else:
-            rng = self.random_state_ # fallback, not-resettable
+        # if seed is not None:
+        #     rng = np.random.default_rng(seed)
+        # else:
+        #     rng = self.random_state_ # fallback, not-resettable
+
+        rng = RandomStateManager.create_rng(self.random_state_)
```
Comment (anneliesdej) [inline src/synthpop/methods/tree_utils.py:165]: I would add some information about the randomness in sample_from_leaves specifically. Every call resets RNG state to the initial seed state. This means that
```python
sampler.sample_from_leaves(A)
sampler.sample_from_leaves(A)
```
Gives the same results. This is what we want to have, but it is not as common. Evolving randomness is more common than replayable randomness.

It might also be enough to make sure this is tested, but I think it is helpful to add as a comment here.

File: src/synthpop/methods/cart_synth.py
Comment (anneliesdej) [inline src/synthpop/methods/cart_synth.py:228]: The default svd_solver of PCA() is 'auto' and it states this:
"The solver is selected by a default ‘auto’ policy is based on X.shape and n_components: if the input data has fewer than 1000 features and more than 10 times as many samples, then the “covariance_eigh” solver is used. Otherwise, if the input data is larger than 500x500 and the number of components to extract is lower than 80% of the smallest dimension of the data, then the more efficient “randomized” method is selected. Otherwise the exact “full” SVD is computed and optionally truncated afterwards."

Which means that randomness can be introduced dependent on the input data. We could force the solver to use 'full', or add a random_state there (probably the preferred option)

File: src/synthpop/methods/sample_synth.py
Comment (anneliesdej) [inline src/synthpop/methods/sample_synth.py]:
```suggestion
```
np is not used

File: src/synthpop/methods/sample_synth.py [hunk 3]
```diff
@@ -92,8 +95,9 @@ def transform(self, X: pd.DataFrame | None) -> pd.Series:
             raise NotFittedError("SampleMethod is not fitted. Call `fit` first.")
         
         n = len(X) if X is not None else self.n_samples_
+        rng = RandomStateManager.create_rng(seed=self.random_state_)
```
Comment (anneliesdej) [inline src/synthpop/methods/sample_synth.py:97]: Maybe also add here that it is a replay RNG and not a streaming RNG

File: tests/unit/test_sample_method.py [hunk 4]
```diff
@@ -92,6 +101,24 @@ def test_transform_values_within_observed_support():
     if any(pd.isna(v) for v in values):
         assert result.isna().any()
 
+def test_transform_uses_random_state_manager(mocker):
```
Comment (anneliesdej) [inline tests/unit/test_sample_method.py:119]: Integration tests need to be added to test reproducibility. Same seed gives same results and repeated calls give same results

File: src/synthpop/methods/sample_synth.py (hunk 3 above)
Comment (KoomenErina) [reply src/synthpop/methods/sample_synth.py:97]: Writing in the docs of `create_rng` that it is a replay RNG would be more efficient I think

File: src/synthpop/methods/tree_utils.py (hunk 2 above)
Comment (KoomenErina) [reply src/synthpop/methods/tree_utils.py:165]: added it to the doc string

File: src/synthpop/methods/sample_synth.py (hunk 3 above)
Comment (anneliesdej) [reply src/synthpop/methods/sample_synth.py:97]: Yes, agreed

File: tests/unit/test_pca_encoder.py [hunk 5]
```diff
@@ -11,9 +11,10 @@
 
 class TransformStub(TransformerMixin, BaseEstimator):
 
-    def __init__(self, fit_return_value=None, transform_return_value=None):
+    def __init__(self, fit_return_value=None, transform_return_value=None,random_state=None):
```
Comment (anneliesdej) [inline tests/unit/test_pca_encoder.py:14]: this file needs an autopep8

File: tests/unit/test_sample_method.py [hunk 6]
```diff
@@ -31,19 +31,28 @@ def make_fitted_model(values: list, counts: list, target_name="target", n_sample
         (pd.Series([np.nan, np.nan, np.nan, np.nan, 0], name="my_target"), [np.nan, 0], [4, 1],),
     ],
 )
-def test_fit_stores_distribution_and_metadata(y, expected_values, expected_counts):
+def test_fit_stores_distribution_and_metadata(y, expected_values, expected_counts,mocker):
```
Comment (anneliesdej) [inline tests/unit/test_sample_method.py]: This file needs an autopep8

File: tests/unit/test_synthesiser.py [hunk 7]
```diff
@@ -174,6 +175,81 @@ def test_synthesiser_fit_custom_order_by_column_index():
...
+
+        def fit(self,X,y):
+            assert MockRandomStateManager.is_in_contextblock, "fit called outside context block"
+            return super().fit(X,y)
+        
+        def transform(self,X):
+            assert MockRandomStateManager.is_in_contextblock, "transform called outside context block"
+            return super().transform(X)
+
+    return mocker.patch(
```
Comment (anneliesdej) [inline tests/unit/test_synthesiser.py:216]: This file needs an autopep8

File: .gitignore [hunk 8]
```diff
@@ -10,5 +10,6 @@
 classes.dot
 pyproject.toml
 tests/unit/__pycache__/
+*.png
```
Comment (anneliesdej) [inline .gitignore]:
```suggestion
```
Is already in the list two lines below

File: tests/unit/test_leaf_node_sampler.py [hunk 9]
```diff
@@ -159,9 +182,12 @@ def integers(self, low, high, size=None):
             out.append(low + (v % (high - low)))
         return np.array(out)
     
-def test_sampling_deterministic_with_stub_rng():
+def test_sampling_deterministic_with_stub_rng(mocker):
+    #TODO: inject the stub by mocking RandomStateManager.create_rng
```
Comment (anneliesdej) [inline tests/unit/test_leaf_node_sampler.py]: Still a todo

File: tests/unit/test_leaf_node_sampler.py [hunk 10]
```diff
@@ -194,6 +220,7 @@ def test_sample_from_leaves_each_missing_type(missing_value):
         assert any(isinstance(v, float) and np.isnan(v) for v in y_syn)
 
 def test_sample_determinism_with_same_seed():
+    #TODO: move to integration test
```
Comment (anneliesdej) [inline tests/unit/test_leaf_node_sampler.py]: still a todo

File: tests/unit/test_sample_method.py [hunk 11]
```diff
@@ -92,6 +101,24 @@ def test_transform_values_within_observed_support():
...
+
+    result = model.transform(pd.DataFrame(index=range(100)))
+
+    mock_create_rng.assert_called_with(seed=123456)
+    mock_sample_array.assert_called_with(expected_rng,model.counts_,model.values_,100)
+    assert (result==expected_result).all()
+
+
 
 def test_transform_reproducibility_with_fixed_seed():
```
Comment (anneliesdej) [inline tests/unit/test_sample_method.py:140]: Here I miss a test that checks what happens with two calls (if it is streaming or replayable)

File: tests/integration/test_int_reproducibility.py [hunk 12]
```diff
@@ -96,3 +99,78 @@ def test_standard_transformer_reproduces_when_setting_seed():
     result2 = transformer1.transform(X=2)
```
Comment (anneliesdej) [inline tests/integration/test_int_reproducibility.py]:
```suggestion
    result2 = transformer2.transform(X=2)
```
I think this should be transfomer2

File: tests/integration/test_int_reproducibility.py [hunk 13]
```diff
@@ -6,8 +6,11 @@
 import numpy as np
 from sklearn.base import BaseEstimator, TransformerMixin
 
+from synthpop.methods.sample_synth import SampleMethod
 from synthpop.reproducibility import RandomStateManager
-
+import pandas as pd
+from synthpop.synthesiser import Synthesiser
+from tests.integration.data_generated_for_tests import get_test_data_classifier,get_test_data_regressor,simulate_realistic_dataset_correlations
```
Comment (anneliesdej) [inline tests/integration/test_int_reproducibility.py:25]: Should the sample method also be added here or are unit tests enough?

File: tests/integration/test_int_reproducibility.py [hunk 14]
```diff
@@ -6,8 +6,11 @@
...
 
 @pytest.fixture(autouse=True)
 def control_random_state_manager():
```
Comment (anneliesdej) [inline tests/integration/test_int_reproducibility.py:27]: Maybe add one line of explanation that we do not use the big fixture as this test needs to set reproducibility and not use the default of the big fixture (in make_int_test_reproducible)

File: tests/integration/test_int_reproducibility.py [hunk 15]
```diff
@@ -96,3 +99,78 @@ def test_standard_transformer_reproduces_when_setting_seed():
...
+        else:
+            d_data[k] = X_clas[k]
+
+    d_data['y1'] = y_reg
+    d_data['y2']= y_clas
+
+    return pd.DataFrame(d_data)
+
+
+
```
Comment (anneliesdej) [inline tests/integration/test_int_reproducibility.py]: I would add one more test for completeness. Right now we verified that
```python
fit(seed)
generate()
generate()
```
is replayable, but I would also add a test that overrides the seed in generate to check if that is also replayable. So something like:
```python
fit(seed)
generate(other seed)
generate(same other seed)
```

File: tests/integration/test_int_reproducibility.py (hunk 13 above)
Comment (KoomenErina) [reply tests/integration/test_int_reproducibility.py:25]: see `test_sample_method_reproducible`, it is already here

File: tests/unit/test_sample_method.py (hunk 11 above)
Comment (KoomenErina) [reply tests/unit/test_sample_method.py:140]: see `test_sample_method_reproducible` in test_int_reproducibility.py

File: tests/unit/test_leaf_node_sampler.py [hunk 16]
```diff
@@ -106,14 +106,37 @@ def test_fit_sampler_sets_y_dtype_correctly(y_input, expected_dtype_check):
 
     assert np.issubdtype(sampler._y_dtype, expected_dtype_check)
 
+def test_fit_sampler_creates_seed_from_random_state_manager(mocker):
```
Comment (anneliesdej) [inline tests/unit/test_leaf_node_sampler.py:115]: This file would benefit from autopep8

File: tests/integration/test_int_tree_method.py [hunk 17]
```diff
@@ -10,9 +10,14 @@
 from synthpop.data_processing.encoders import PCAEncoder
 from synthpop.data_processing.missing_value_handling import MissingValuePredictor
 from synthpop.methods.cart_synth import TreeClassifierMethod, TreeRegressorMethod
+from synthpop.reproducibility import RandomStateManager
 from synthpop.utils import str_dtype
 from sklearn.datasets import make_classification, make_regression
```
Comment (anneliesdej) [inline tests/integration/test_int_tree_method.py:15]: This file would benefit from autopep8

File: tests/integration/test_int_reproducibility.py [hunk 18]
```diff
@@ -93,6 +103,119 @@ def test_standard_transformer_reproduces_when_setting_seed():
...
+def combined_regressor_and_classifier_test_data(seed=10):
+    X_reg, y_reg = get_test_data_regressor(
+        seed=seed, with_cats=True, with_missing_features=True, with_missing_target=True)
+    X_clas, y_clas = get_test_data_classifier(
+        seed=seed, with_cats=True, with_missing_features=True, with_missing_target=True)
+
+    d_data = {}
+
+    available_columns = set(X_reg.keys()).intersection(set(X_clas.keys()))
```
Comment (anneliesdej) [inline tests/integration/test_int_reproducibility.py]:
```suggestion
    available_columns = sorted(set(X_reg) & set(X_clas))
```
The helper currently is technically non-deterministic as python sets are unordered. Since it is used for reproducibility tests, I'd make it deterministic

File: tests/integration/test_int_reproducibility.py [hunk 19]
```diff
@@ -93,6 +103,119 @@ def test_standard_transformer_reproduces_when_setting_seed():
...
+
+    syn1 = synth.generate()
+    syn2 = synth.generate(random_seed=100)
+    syn3 = synth.generate(random_seed=100)
+
+    assert syn2.equals(syn3)
+    assert not syn2.equals(syn1)
+
+
+def test_sample_method_reproducible():
```
Comment (anneliesdej) [inline tests/integration/test_int_reproducibility.py:201]: For completeness I would not only check if transform is replayable but also if a seed is reproducible (like the other tests have):
```python
def test_sample_method_explicit_seed_reproducible():

    y = ...

    m1 = SampleMethod(random_state=123).fit(None, y)
    m2 = SampleMethod(random_state=123).fit(None, y)

    assert m1.transform(None).equals(m2.transform(None))
```

File: tests/integration/test_int_reproducibility.py [hunk 20]
```diff
@@ -93,6 +103,119 @@ def test_standard_transformer_reproduces_when_setting_seed():
...
+    syn2 = synth.generate(n=100, random_seed=1234)
+
+    assert not syn1.equals(syn2)
+
+    synth2 = Synthesiser(random_seed=1234)
+    syn3 = synth2.fit(obs).generate(n=100)
+    assert not syn1.equals(syn3)
+
+
+def test_generate_independent_syn_datasets_reproducible():
```
Comment (anneliesdej) [inline tests/integration/test_int_reproducibility.py]:
```suggestion
def test_generate_override_seed_is_replayable():
```
This describes the test better

File: src/synthpop/data_processing/missing_value_handling.py [hunk 21]
```diff
@@ -140,7 +141,7 @@ def prepare_data_for_fit(self, X: Dict[str, npt.NDArray], y: npt.NDArray) -> tup
 
         self.feature_order_ = list(X_val.keys())
```
Comment (anneliesdej) [inline src/synthpop/data_processing/missing_value_handling.py:143]: This file would benefit from autopep8

File: src/synthpop/methods/tree_utils.py [hunk 22]
```diff
@@ -159,12 +161,14 @@ def sample_from_leaves(self, leaf_ids: npt.ArrayLike) -> npt.NDArray:
         ):
             raise NotFittedError("LeafNodeSampler is not fitted. Call `fit_sampler` first.")
         
-        seed = getattr(self, "_seed", None)
+        # seed = getattr(self, "_seed", None)
```
Comment (anneliesdej) [inline src/synthpop/methods/tree_utils.py]: I think these comment lines can be removed

File: src/synthpop/reproducibility.py
Comment (casvanerpduo) [inline src/synthpop/reproducibility.py:28]: Also add an Example on how to use this with the Synthesizer class and synthesize data. That is now unclear to me.

Perhaps also extend the doc strings to also explain how it manages reproducibility, and why that is required

Comment (KoomenErina) [reply src/synthpop/reproducibility.py:28]: The user does not use this class directly when using the Synthesiser class. Examples of reproducible synthesis are in the example docs

Comment (KoomenErina) [reply src/synthpop/reproducibility.py:28]: The functional description (reproducibility.md) explains the why, and for a part the how. The developerguide (randomness.md) explains it more practically

File: tests/integration/data_generated_for_tests.py [hunk 23]
```diff
@@ -0,0 +1,111 @@
+import copy
+import pytest
+import pandas as pd
+import numpy as np
+import string
+
+from synthpop.utils import str_dtype
+from sklearn.datasets import make_classification, make_regression
```
Comment (casvanerpduo) [inline tests/integration/data_generated_for_tests.py]:
```suggestion
import copy
import string

import pytest
import pandas as pd
import numpy as np
from sklearn.datasets import make_classification, make_regression

from synthpop.utils import str_dtype
```

File: tests/integration/test_int_reproducibility.py [hunk 24]
```diff
@@ -6,11 +6,21 @@
 import numpy as np
 from sklearn.base import BaseEstimator, TransformerMixin
 
+from synthpop.methods.sample_synth import SampleMethod
+from synthpop.methods.tree_utils import LeafNodeSampler
 from synthpop.reproducibility import RandomStateManager
+import pandas as pd
+from synthpop.synthesiser import Synthesiser
+from tests.integration.data_generated_for_tests import get_test_data_classifier, get_test_data_regressor, simulate_realistic_dataset_correlations
```
Comment (casvanerpduo) [inline tests/integration/test_int_reproducibility.py:23]:
```suggestion
import pandas as pd
import pytest
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin

from synthpop.methods.sample_synth import SampleMethod
from synthpop.methods.tree_utils import LeafNodeSampler
from synthpop.reproducibility import RandomStateManager
from synthpop.synthesiser import Synthesiser
from tests.integration.data_generated_for_tests import get_test_data_classifier, get_test_data_regressor, simulate_realistic_dataset_correlations
```
